lib/events/api.go

```diff
@@ -799,10 +799,14 @@ type UploadMetadataGetter interface {
 	GetUploadMetadata(sid session.ID) UploadMetadata
 }
 
-// StreamWriter implements io.Writer to be plugged into the multi-writer
-// associated with every session. It forwards session stream to the audit log
-type StreamWriter interface {
+// SessionRecorder records session events. It can be used both as a
+// [io.Writer] when recording raw session data and as a [apievents.Recorder]
+// when recording session events.
+type SessionRecorder interface {
 	io.Writer
+	// SetupEvent will set necessary event fields, and must be called before
+	// the event is recorded.
+	SetupEvent(event apievents.AuditEvent) error
```

Comment:
Does this have to be in the recorder interface?

With this, each recorder, whether it be a streaming recorder, one that writes to disk, a test one that just saves a copy in memory, or any future implementation is forced to implement the same logic for setting necessary fields.

Can we clean this up by either:
- moving the event setup upstream, so the recorder just records, and whatever calls the recorder is responsible for providing valid data
- wrapping with a checking streamer (or whatever its called now) in order to centralize this logic

Reply:
The way it's implemented now, `SessionWriter` sets event indexes which requires keeping state. This code could be split from `SessionWriter`, but as we only have 2 recorders now that we have been exclusively for years (`SessionWriter` and `DiscardRecorder`), I think it's better to cross this bridge if we get to it.

Reply:
The reason I split `SetupEvent` is because previously, `lib/events.SessionWriter` modified events in the caller's goroutine (called SetupEvent) and `lib/events.AsyncEmitter` read the same fields of the same events in a different goroutine, so events had to be recorded before they were emitted if doing both was desired to prevent races. These 2 structs are always used to record and emit respectively in production.

So centralizing this logic would be nice, but would open us up to easily adding races in the future by accident.

Reply:
I still think we can fix the race without requiring a bunch of independent implementations of `SetupEvent`. Let me know if you want to brainstorm some solutions here,.

Reply:
Updated to have a separate interface for setting up events.